src/main/scala/io/iohk/ethereum/blockchain/sync/FastSync.scala

```diff
@@ -768,12 +429,15 @@ object FastSync {
 
     def blockChainWorkQueued: Boolean =  blockBodiesQueue.nonEmpty || receiptsQueue.nonEmpty
 
-    def updateNextBlockToValidate(header: BlockHeader, K: Int, X:Int): SyncState = copy(
+    def blockChainWorkLeft: Boolean =
+      bestBlockHeaderNumber < safeDownloadTarget || blockChainWorkQueued
+
+    def updateNextBlockToValidate(header: BlockHeader, K: Int, X:Int, randomFun: Int => Int): SyncState = copy(
```

Comment:
missing space between `X` and `:` also you're using only `header.number` from header